src/python/bot/tasks/train_gradientfuzz_task.py

```diff
@@ -13,9 +13,251 @@
+
+
+def get_corpus_directory(root_dir, fuzzer_name):
+  """Gets corpus directory path based on fuzzer."""
+  return os.path.join(root_dir, run_constants.CORPUS_DIR,
+                      fuzzer_name + run_constants.CORPUS_SUFFIX)
+
+
+def upload_model_to_gcs(model_directory, fuzzer_name):
+  """
+  Upload the entire model directory to GCS bucket.
+
+  Note that metadata stored within the model directory is needed
+  for the generation scripts.
+
+  Args:
+    model_directory (str): models/[architecture]/[run-name].
+    fuzzer_name (str): The fuzzer the model is trained for.
+
+  Returns:
+    True if corpus can be acquired and False otherwise.
+  """
+  # Get GCS model path.
+  gcs_model_directory = ml_train_utils.get_gcs_model_directory(
+      run_constants.GRADIENTFUZZ_DIR)
+  if not gcs_model_directory:
+    logs.log_error('Failed to upload model: cannot get GCS model bucket.')
+    return
+
+  # Zip entire model directory and upload.
+  model_dir_name = os.path.basename(model_directory)
+  zipped_dir = shutil.make_archive(model_dir_name, 'zip', model_directory)
+  gcs_model_path = '{}/{}'.format(gcs_model_directory, zipped_dir)
+
+  logs.log('Uploading the model for fuzzer {} and run {} to {}.'.format(
+      fuzzer_name, model_dir_name, gcs_model_path))
+
+  # Upload files to GCS.
+  result = (storage.copy_file_to(zipped_dir, gcs_model_path))
+
+  if result:
+    logs.log('Uploaded GradientFuzz model {} for fuzzer {}.'.format(
+        model_dir_name, fuzzer_name))
+  else:
+    logs.log_error(
+        'Failed to upload GradientFuzz model {} for fuzzer {}.'.format(
+            model_dir_name, fuzzer_name))
+
+
+def gen_inputs_labels(corpus_directory, fuzzer_binary_path):
+  """
+  Generates inputs and labels from raw input corpus.
+
+  Args:
+    corpus_directory (str): Path to raw inputs.
+    fuzzer_binary_path (str): Path to compiled fuzz target binary.
+
+  Returns:
+    (new_process.ProcessResult): Result of `run_and_wait()`.
+    (str): Dataset name (results stored under
+        GRADIENTFUZZ_SCRIPTS_DIR/data/[dataset_name]).
+  """
+  script_path = get_script_path(run_constants.GENERATE_DATA_SCRIPT)
+  dataset_name = os.path.basename(corpus_directory)
+  args_list = [
+      script_path,
+      run_constants.FUZZ_TARGET_BINARY_FLAG,
+      fuzzer_binary_path,
+      run_constants.INPUT_DIR_FLAG,
+      corpus_directory,
+      run_constants.DATASET_NAME_FLAG,
+      dataset_name,
+      run_constants.MEDIAN_MULT_FLAG,
+      run_constants.DEFAULT_MEDIAN_MULT_CUTOFF,
+  ]
+
+  logs.log('Launching input gen with args: "{}".'.format(str(args_list)))
+
+  # Run process in GradientFuzz directory.
+  data_gen_proc = new_process.ProcessRunner(sys.executable)
+
+  return data_gen_proc.run_and_wait(
+      additional_args=args_list,
+      cwd=GRADIENTFUZZ_SCRIPTS_DIR,
+      timeout=run_constants.DATA_GEN_TIMEOUT), dataset_name
+
+
+def train_gradientfuzz(fuzzer_name, dataset_name, num_inputs, testing):
```

Comment:
How big of a deal would it be to remove the testing argument? I don't really like the idea of the tests complicating the implementation. If it really is necessary, I'd prefer something like:

batch_size = environment.get_value('GRADIENTFUZZ_BATCH_SIZE', min(32, int(num_inputs * 0.4)))

with the environment variable set for testing, or at minimum defaulting to false for testing here to simplify calls.

Reply:
Got it -- unfortunately, since the sample corpus I'm using for testing is pretty small we need small batches or the training doesn't work at all :( I'll try the environment variable option.